### PR TITLE
added 'build' subcmd to __main__.py

### PR DESCRIPTION
**Pull request by shakfu**
This change adds a `build` subcmd to `__main__.py` and a `zig_build_config function` in `buildzig.py` for configuring builds from the command line (i.e. without using pyproject.toml).

This essentially this enables the following:

```bash
virtualenv venv && source venv/bin/activate
git clone https://github.com/shakfu/ziggy-pydust
cd ziggy-pydust
pip install -e .
cd ../demo # see attached demo
python -m pydust build --ext-name 'fib._lib' --ext-path 'fib/fib.zig'
```

Note that although the extension is successfully built and works via direct testing, at the end of the build an error/warning is output: "Not a valid python module, skipping... fib._lib No module named 'fib'"

[demo.zip](https://github.com/fulcrum-so/ziggy-pydust/files/12777619/demo.zip)

### CODE REVIEW

**Comment by shakfu** [comment]
> Note that although the extension is successfully built and can be tested as such, at the end of build an error/warning is output: "Not a valid python module, skipping... fib._lib No module named 'fib'"

This issue is resolved if `generate_stubs.py` is copied into the containing folder: 

```bash
python generate_stubs.py fib._lib .
```
which generates `_lib.pyi` in `fib`:
```
from __future__ import annotations

def nth_fibonacci_iterative(n, /): ...
def nth_fibonacci_recursive(n, /, *, f0=0, f1=1): ...
def nth_fibonacci_recursive_tail(n, /): ...

class Fibonacci:
    def __init__(first_n, /):
        pass
    def __iter__(self, /):
        """
        Implement iter(self).
        """
        ...

class FibonacciIterator:
    def __init__(i, ith, next, stop, /):
        pass
    def __next__(self, /):
        """
        Implement next(self).
        """
        ...
```

**Comment by shakfu** [comment]
Added `generate_stubs` options to to `build` subcmd in `__main__.py` such that this works now:

```bash
python -m pydust build --generate-stubs --ext-name 'fib._lib' --ext-path 'fib/fib.zig'
```

**Comment by shakfu** [comment]
Added a `Builder` class in `__init__.py` to serve as a frontend class for both programmatic usage and also command line usage.

**Comment by gatesn** [comment]
Ah this is quite neat, thank you! 

Do you think that this could be restructured around the current config object though? By that I mean, leave the arg parsing in __main__.py, and with the build subcommand, populate a ToolPydust object and pass that into zig_build.

So `def zig_build(argv: list[str]):` turns into `def zig_build(argv: list[str], config: ToolPydust):`

**Comment by shakfu** [comment]
> Do you think that this could be restructured around the current config object though? By that I mean, leave the arg parsing in **main**.py, and with the build subcommand, populate a ToolPydust object and pass that into zig_build.
> 
> So `def zig_build(argv: list[str]):` turns into `def zig_build(argv: list[str], config: ToolPydust):`

Sure. While I personally prefer the class with command line api of the penultimate commit, I appreciate you want to keep the current functional api.

The latest commit provides the api as per your request, such that `zig_build` is now `def zig_build(argv: list[str], conf: config.ToolPydust | None)`

File: pydust/buildzig.py [hunk 1]
```diff
@@ -34,25 +34,25 @@
 PYLDLIB = os.path.splitext(PYLDLIB)[0]
 
 
-def zig_build(argv: list[str]):
-    conf = config.load()
+def zig_build(argv: list[str], conf: config.ToolPydust | None):
+    if not conf:
+        conf = config.load()
```
Comment (gatesn) [inline pydust/buildzig.py]:
```suggestion
    conf = conf or config.load()
```

File: pydust/__main__.py [hunk 2]
```diff
@@ -12,15 +12,40 @@
 limitations under the License.
 """
 
+"""
```
Comment (gatesn) [inline pydust/__main__.py]: Our copyright plugin screwed up

File: pydust/buildzig.py (hunk 1 above)
Comment (shakfu) [reply pydust/buildzig.py]: Done. Thanks for the review.

File: pydust/__main__.py (hunk 2 above)
Comment (shakfu) [reply pydust/__main__.py]: Done.

**Comment by gatesn** [comment]
Ah sorry, you got busted by Ruff

**Comment by shakfu** [comment]
> Ah sorry, you got busted by Ruff

Ok that's fixed :smile:

**Review by gatesn — APPROVED** [review]

**Comment by shakfu** [comment]
@gatesn I'm still getting a CI build failure.  Could this be due to the stub_generation issue highlighted earlier?

**Comment by gatesn** [comment]
Hmm, are you unable to see the output of CI? https://github.com/fulcrum-so/ziggy-pydust/actions/runs/6381967387/job/17319854114?pr=158

Looks like you're missing a  `conf` argument in `ziggy-pydust/test/conftest.py:23`

**Comment by shakfu** [comment]
> Hmm, are you unable to see the output of CI? https://github.com/fulcrum-so/ziggy-pydust/actions/runs/6381967387/job/17319854114?pr=158
> 
> Looks like you're missing a `conf` argument in `ziggy-pydust/test/conftest.py:23`

Thanks, I'll have a look at fixing this.

**Comment by shakfu** [comment]
@gatesn 

The issue was this `def zig_build(argv: list[str], conf: config.ToolPydust | None)` instead of `def zig_build(argv: list[str], conf: config.ToolPydust | None = None)`.

That's fixed in the latest commit

**Comment by shakfu** [comment]
@gatesn Thanks for the merge!

**Pull request merged**